Implementation of Peace Implants

imw-Oculus opened the pull request:
New feature: Peace implants
- Functions by using a new part of the is_pacified() proc
- This can be suppressed by having >= Methylin in your system
- Drains a small amount of nutrition from the host
- Destroyed if removed from host

Requesting:
- Balance input
- Cargo cost
- Placed by default in the Warden's locker?
- Research barrier?

:cl:
* rscadd: Adds peace (pax) implants. They steadily feed off of the host's nutrition, and prevent the host from performing violent actions. Temporarily malfunctions if the host has >=15u Methylin in their system, or is starving. Implant is destroyed upon removal from the host. They can be found in the Trader Vendor, for now.

Probe1 commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -293,6 +298,8 @@ the implant may become unstable and either pre-maturely inject the subject or si
 	return 1
 
+
+//Greytide Implant
```
All of these comments are superfluous.  Comments should explain code.  I immediately know that this is a graytide implant because the name is right there ``` name = "Greytide Implant" ``` .  This is applicable on all of your comments naming objects that are immediately named but I'm not going to leave a comment on all of them.

imw-Oculus replied:
Noted: I'll fix that tomorrow.

imw-Oculus commented:
Noted: I'll take care of that tomorrow.

Probe1 commented:
Don't get me wrong we have _plenty_ of funny comments that aren't strictly needed.  But adding a shitload of comments to tell people what they already know just results in bloated files and more shit for coders to scan past.  I made this same mistake when I started.

imw-Oculus commented:
No worries, I'm striving to find the correct use for comments with this project. I usually comment the ever loving shit out of my code, because I'll drop projects and come back to them six months later, so it's taking a little getting used to.

SarahJohnson489 commented:
Like the chem implant, have it start empty so they have to work with the trader to get it.

Probe1 commented:
Here's a very useful article on our (very standard) practice for commenting and such.
https://en.wikipedia.org/wiki/Self-documenting_code .  We avoid magic numbers and nondescript variable names so our code is _self documenting_ and doesn't require comments to explain what it is unless it's utilized in an obscure way or complicated to understand.  

Here is a well commented file that borders on being too explanatory but is acceptable.  You can read this and right away know how it works.  It is a good example of commenting practices.
https://github.com/vgstation-coders/vgstation13/blob/Bleeding-Edge/code/modules/randomMaps/vaults.dm

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+/obj/item/weapon/implant/peace/process()
+	var/mob/living/carbon/host = imp_in
+	if (isnull(host))
+		if (imp_alive == 1)
```
This can only be 0 or 1, you can just check for `(imp_alive)`, but it's up to you.

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+	var/mob/living/carbon/host = imp_in
+	if (isnull(host))
+		if (imp_alive == 1)
+			Destroy()
```
`Destroy` is supposed to be called by the garbage collector, you don't want to call it manually. Instead, you want to `qdel(src)` if your intent is for the implant to delete itself, or just have it remove itself from `processing_list` if that's what you wanted.

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+			Destroy()
+			return 0
+	else
+		if (imp_alive == 0)
```
This can only be 0 or 1, you can just check for `(!imp_alive)`, but it's up to you

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+		if (imp_alive == 0)
+			imp_alive = 1
+		if (!host.reagents.has_reagent(CHILLWAX,1) && !host.reagents.has_reagent(ANTI_TOXIN))
+			if (msg_debounce == 1)
```
this can only be 0 or 1, you can just check for `(msg_debounce)` but it's up to you.

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+/obj/item/weapon/implant/peace/Destroy()
+	processing_objects.Remove(src)
+	..()
+	return
```
This is a trailing return, it does nothing. Remove it.

DamianX commented on code/game/objects/items/weapons/implants/implantcase.dm:
```diff
@@ -121,3 +121,15 @@
+
+//Pax(Peace) Implant Case
+/obj/item/weapon/implantcase/peace
+	name = "Glass Case- 'Pax'"
```
The capitalization here is incorrect, it should read `glass case - 'Pax'`, otherwise it gets treated as a proper noun.

DamianX commented on code/game/objects/items/weapons/implants/implantcase.dm:
```diff
@@ -121,3 +121,15 @@
+	New()
+		src.imp = new /obj/item/weapon/implant/peace(src)
+		..()
+		return
```
trailing return, delet

DamianX commented on code/game/objects/items/weapons/implants/implantcase.dm:
```diff
@@ -121,3 +121,15 @@
+	icon = 'icons/obj/items.dmi'
+	icon_state = "implantcase-b"
+
+	New()
```
this is a relative path, also known as the devil. It's one of the few things we have guidelines for https://github.com/vgstation-coders/vgstation13/wiki/Coding-standards-for-vgstation13#absolute-pathing

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+	var/imp_alive = 0
+	var/msg_debounce = 0
+	var/imp_data = {"
+<b>Implant Specifications:</b><BR>
```
Why Do People Type Like This? I Am Puzzled

imw-Oculus replied:
Because it's the title of the device's display page (I think?)

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+	if (isnull(host))
+		if (imp_alive == 1)
+			Destroy()
+			return 0
```
This return value is ignored, no reason not to just `return`

imw-Oculus replied:
Thanks for the input Dam, I'll clean this up tomorrow. I'll take the style pointers to heart as well.

Durgen commented:
Making these easily available would render chiliwax farming pointless. Perhaps a lockbox could be added to the tradevend stock instead of the warden’s locker or RnD. This way traders would keep their monopoly on the reagent.

DrCelt commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,56 @@ the implant may become unstable and either pre-maturely inject the subject or si
+				msg_debounce = 0
+				to_chat(host, "<span class = 'notice'>Your rage bubbles, your Pax implant is no longer suppressed!</span>")
+			host.reagents.add_reagent(CHILLWAX,2)
+		if (host.reagents.has_reagent(ANTI_TOXIN) && msg_debounce == 0)
```
&& !msg_debounce

imw-Oculus commented:
@Durgen Someone else requested this, so I coded it. I dunno. I'd argue that you'd have to farm an inordinate amount of chill wax and frequently refuel the implant if you wanted to have the implant run on farmed chill wax. I'm open to balance thoughts, what are yours?

Dacendeth commented:
You could make it a trader item, box of chillwax implants but don't include the implanter tool so the traders cant use it themselves without some legwork. That way it makes it so that it's more of an item to sell to security.

jknpj commented:
>You could make it a trader item

Yes let give antag traders more rape dungeon toys

MadmanMartian commented:
>Yes let give antag traders more rape dungeon toys
>> but don't include the implanter tool so the traders cant use it themselves

jknpj commented:
I stopped reading halfway in a blind moment of rage :sob:

imw-Oculus commented:
Okey dokey. Removed comments, removed relative pathing, cleaned up my boolean conditional statements. Compiling and testing. Also going to open a separate PR when this is done to clean up the relative pathing in implants, and some naming conventions - along the lines that Durge suggested.

imw-Oculus commented:
Hang on. Hunger isn't causing the implant to deactivate. Sigh. Stand-by.

Durgen commented:
Dacendeth’s idea is good.

MadmanMartian commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,77 @@ the implant may become unstable and either pre-maturely inject the subject or si
+"}
+
+/obj/item/weapon/implant/peace/meltdown()
+	for (var/mob/msg_viewer in viewers(src, null))
```
This is the old way of doing visible_message. Use visible_message unless you've got a good reason to use this.

MadmanMartian commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,77 @@ the implant may become unstable and either pre-maturely inject the subject or si
+
+/obj/item/weapon/implant/peace/meltdown()
+	for (var/mob/msg_viewer in viewers(src, null))
+		msg_viewer.show_message("<span class='warning'>The [name] releases a dying hiss as it denatures!</span>", 1)
```
\The [src] instead of The [name]. The ToString of an object just returns its name anyway.

MadmanMartian commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,77 @@ the implant may become unstable and either pre-maturely inject the subject or si
+
+		if (!imp_msg_debounce && malfunction == MALFUNCTION_TEMPORARY)
+			imp_msg_debounce = 1
+			to_chat(host, "<span class = 'warning'>Your rage bubbles, the [name] inside you is being suppressed!</span>")
```
\the [src] instead of the [name]

MadmanMartian commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,77 @@ the implant may become unstable and either pre-maturely inject the subject or si
+
+		if (imp_msg_debounce && !malfunction)
+			imp_msg_debounce = 0
+			to_chat(host, "<span class = 'warning'>Your rage cools, the [name] inside you is active!</span>")
```
\the [src] instead of the [name]

imw-Oculus commented:
Wew.
Compiling and testing:
- A couple of small bug fixes
- Syntax fixes proposed by Martian
- Adding implant cases to the Vox vendor (but no implanters)

imw-Oculus commented:
Final tests complete:
- Implanting
- Implant removal
- Implant function
- Vendor item

DamianX commented on code/game/objects/items/weapons/implants/implant.dm:
```diff
@@ -520,6 +529,73 @@ the implant may become unstable and either pre-maturely inject the subject or si
+"}
+
+/obj/item/weapon/implant/peace/meltdown()
+	visible_message("<span class='warning'>\The [src] releases a dying hiss as it denatures!</span>", 1)
```
There's a holdover `1`

imw-Oculus replied:
Thanks for catching that - updooting

Blithering commented:
Bare in mind traitor birds could buy the cheapest implant (freedom I think) and reuse that implanter in their bid to rob hardworking station valid hunters of their right to commit acts of unspeakable violence
